crux-lucene/src/crux/lucene.clj

```diff
@@ -174,79 +172,142 @@
                         ;; Used for wildcard searches
-                        (.add (StringField. field-crux-attr, (keyword->k k), Field$Store/YES)))]]
+                        (.add (StringField. field-crux-attr, (keyword->k a), Field$Store/YES)))]]
       (.updateDocument ^IndexWriter index-writer (Term. field-crux-id id-str) doc)))
 
   (evict! [_ index-writer eids]
-    (with-open [index-snapshot (db/open-index-snapshot index-store)]
-      (let [attrs-id->attr (->> (db/all-attrs index-snapshot)
-                                (map #(vector (->hash-str %) %))
-                                (into {}))
-            qs (for [[a v] (db/exclusive-avs index-store eids)
-                     :let [a (attrs-id->attr (->hash-str a))
-                           v (db/decode-value index-snapshot v)]
-                     :when (not= :crux.db/id a)]
-                 (TermQuery. (Term. field-crux-id (->hash-str (DocumentId. a v)))))]
-        (.deleteDocuments ^IndexWriter index-writer ^"[Lorg.apache.lucene.search.Query;" (into-array Query qs))))))
-
-(defn ->indexer
-  {::sys/deps {:index-store :crux/index-store}}
-  [{:keys [index-store]}]
-  (LuceneAvIndexer. index-store))
-
-(defn ->analyzer
-  [_]
+    (let [qs (for [eid eids]
+               (TermQuery. (Term. field-crux-eid (->hash-str eid))))]
+      (.deleteDocuments ^IndexWriter index-writer ^"[Lorg.apache.lucene.search.Query;" (into-array Query qs)))))
+
+(defn ->indexer [_]
+  (LuceneAveIndexer.))
+
+(defn ->analyzer [_]
   (StandardAnalyzer.))
 
+(defn- transform-tx-events [document-store tx-events]
+  (let [conformed-tx-events (map txc/<-tx-event tx-events)
+        docs (db/fetch-docs document-store
+                            (txc/conformed-tx-events->doc-hashes conformed-tx-events))]
+    (reduce (fn [acc {:keys [op] :as tx-event}]
+              (case op
+                :crux.tx/evict (-> acc (update :evicted-eids conj (:eid tx-event)))
+                :crux.tx/fn (let [{:keys [args-content-hash]} tx-event
+                                  ;; doc replaced by this point
+                                  nested-events (-> (db/fetch-docs document-store #{args-content-hash})
+                                                    (get args-content-hash)
+                                                    :crux.db.fn/tx-events)
+                                  {:keys [docs evicted-eids]} (transform-tx-events document-store nested-events)]
+                              {:docs (into (:docs acc) docs)
+                               :evicted-eids (into (:evicted-eids acc) evicted-eids)})
+                acc))
+            {:docs docs
+             :evicted-eids #{}}
+            conformed-tx-events)))
+
+(defn- checkpoint-src [^IndexWriter index-writer]
+  (let [^SnapshotDeletionPolicy snapshotter (.getIndexDeletionPolicy (.getConfig index-writer))]
+    (reify cp/CheckpointSource
+      (save-checkpoint [_ dir]
+        (.commit index-writer) ; RFC: do we need this?
```

Comment:
I think so! Based on "When you make a snapshot, it grabs the current commit point and holds a reference to it" via https://freecontent.manning.com/hot-backups-with-lucene/

Reply:
It does, but we also have the option of using the last commit point from however long ago the schedule ran.

I quite like having this `commit` here as it means the checkpoint is as up-to-date as it can be - the question is whether it's superfluous, or even hinders if the schedule runs at the same time.

Reply:
Agreed, the commit frequency and the snapshot frequency feel like they should stay unrelated, and therefore it seems the least surprising option for the user.